osu.Game/Screens/Play/MasterGameplayClockContainer.cs

```diff
@@ -216,11 +216,25 @@ private class HardwareCorrectionOffsetClock : FramedOffsetClock
         {
             // we always want to apply the same real-time offset, so it should be adjusted by the difference in playback rate (from realtime) to achieve this.
             // base implementation already adds offset at 1.0 rate, so we only add the difference from that here.
-            public override double CurrentTime => base.CurrentTime + Offset * (Rate - 1);
+            public override double CurrentTime => base.CurrentTime + offsetAdjust;
 
-            public HardwareCorrectionOffsetClock(IClock source, bool processSource = true)
-                : base(source, processSource)
+            private readonly BindableDouble pauseRateAdjust;
+
+            private double offsetAdjust;
+
+            public HardwareCorrectionOffsetClock(IClock source, BindableDouble pauseRateAdjust)
+                : base(source)
+            {
+                this.pauseRateAdjust = pauseRateAdjust;
+            }
+
+            public override void ProcessFrame()
             {
+                base.ProcessFrame();
+
+                // changing this during the pause transform effect will cause a potentially large offset to be suddenly applied as we approach zero rate.
+                if (pauseRateAdjust.Value == 1)
+                    offsetAdjust = Offset * (Rate - 1);
```

Comment:
The only case I'm afraid of with this way is if `CurrentTime` gets accessed before the first call to `ProcessFrame`, where seeking to a certain position would apply the offset inversely in `MasterGameplayClockContainer.Seek()` but would still not have `CurrentTime` correct due to `offsetAdjust` still being zero.

That should theortically be the case in `MultiSpectatorScreen`, as it stops the clock before the first `Update` call to the gameplay clock container where the clock would receive the first `ProcessFrame` (and even if it gets it somehow, it will still be blocked by `pauseRateAdjust.Value` already being not one):
https://github.com/ppy/osu/blob/9bacc0d50cae548b89341e14507c0c9e54189966/osu.Game/Screens/OnlinePlay/Multiplayer/Spectate/MultiSpectatorScreen.cs#L150-L159

Changing `Seek` so that it also has this `offsetAdjust` logic might be one way to solve this, however.

Reply:
Has this issue been forgotten?

Reply:
Not forgotten, but I'm also not sure if valid.. I'll need to read this comment multiple more times to understand the implications. Having a test scene covering the issue/regression would definitely be helpful.

Reply:
I didn't forget about this review comment. I just don't understand what it's trying to say or what a practical scenario for reproducing such a failure would be.

Reply:
Coming back to this, it turns out the point I brought isn't actually correct. To put it in code rather than words:
```csharp
Add(clock = new MasterGameplayClockContainer());

// ensure gameplay clock in stopped state so offsetAdjust doesn't get updated.
clock.Stop();
clock.Reset();

// set arbitrary offset
config.SetValue(OsuSetting.AudioOffset, 200.0);

clock.Seek(2500);
// initial thought: CurrentTime = 2300
// actual and correct: CurrentTime = 2500
clock.Seek(5000);
// initial thought: CurrentTime = 4800
// actual and correct: CurrentTime = 5000
```

What I was thinking at the point of https://github.com/ppy/osu/pull/14776#discussion_r710860138 is that when the GCC is stopped, the `Seek` calls will get [incorrectly adjusted by the offset](https://github.com/ppy/osu/blob/5f27f1c099c0b1e369cf01662bcf758de91ae42d/osu.Game/Screens/Play/MasterGameplayClockContainer.cs#L131) and the `CurrentTime` will be decreased by the offset from the requested time, as the `HardwareCorrectionOffsetClock` will never update `offsetAdjust`.

But what I didn't realize is that when `offsetAdjust` is never updated (i.e. `offsetAdjust = 0`), the offset will still be applied to the current time as if `Rate` is 1, rather than not applying anything (perhaps the naming is what threw me off in the first place 😛).